### ISSUE TITLE
Please provide an AppImage for download

### ISSUE THREAD

**Issue by probonopd** [notepadqq/notepadqq#817]
Seems like the AppImage was removed and is no longer available?

https://github.com/notepadqq/notepadqq/commit/3eed7e8892d437028f2d362e41c1c83a403d3853

Why?

Please provide an AppImage for download.

**Comment by danieleds** [comment]
Hey, I removed it since we had another packaging mechanism (snap) and keeping many different packages was difficult to maintain at the time.

If you want to bring it back in a separate repo, I'll be very happy to link to it from the README as an alternative installation method.

**Comment by probonopd** [comment]
Hi @danieleds was it creating additional work? Once it's automated, it should normally not require extra work anymore. As per the AppImage philosophy, we encourage upstream application authors to provide official builds (rather than "random" third-party ones).

**Comment by danieleds** [comment]
@probonopd This project has just a few active contributors, so I want to avoid complexity as much as possible in order to let all the work be where it needs to be, i.e. on the core code. AppImage (like any other packaging system) needs work to keep it updated with the changes in the build pipeline or in the project dependencies.

Still, I'm aware that AppImage is something highly requested and useful.

Would you (or someone else) be willing to maintain an official AppImage in a separate repository within the official notepadqq organization?

**Comment by danieleds** [comment]
Also, I don't know if it is possible, but it would be really cool to have an automated way to build and publish both continuous "unstable" AppImages and stable AppImages based on whether that commit is associated with a version tag or not.

**Comment by probonopd** [comment]
> Would you (or someone else) be willing to maintain an official AppImage in a separate repository within the official notepadqq organization?

Putting it into a separate directory directly contradicts this:

> Also, I don't know if it is possible, but it would be really cool to have an automated way to build and publish both continuous "unstable" AppImages and stable AppImages based on whether that commit is associated with a version tag or not.

At least from what I know, the integration with GitHub only works if the builds are done in the same repository in which the code lives. This will not only give you continuous builds, but also builds for pull requests (which can be uploaded to some temporary location only displayed in the build log rather than to GitHub Releases).

**Comment by danieleds** [comment]
Okay. Can it still be in the main repo but somewhat self-contained in its own subdirectory?
The implementation we had was a bit interwoven with travis.yml. I guess it would be enough to just call an external script from travis?

Basically what I'm looking is just something that would provide

* automated "edge" (pre-release) builds and uploads at each commit.
* automated "stable" builds and uploads at each "v*" tag
* the edge appimage should be self-updatable with new versions from the edge channel
* the stable appimage should be self-updatable with new versions from the stable channel
* most declarative as possible, less code as possible

Would you have time to provide some basic working structure for this?

**Comment by probonopd** [comment]
> The implementation we had was a bit interwoven with travis.yml. I guess it would be enough to just call an external script from travis?

Yes, this is possible. I can write a script that does the AppImage generation. As you say, the script needs to be invoked by `.travis.yml`.

> the edge appimage should be self-updatable with new versions from the edge channel
> the stable appimage should be self-updatable with new versions from the stable channel

Let me re-phrase your requirement to see if we have the same understanding here:
Continuous builds (pre-release) should update to continuous builds, whereas released (tagged) builds should update to releases (tagged) builds.

This is definitely something I want to support, even though I am not entirely sure whether some is still necessary in the `linuxdeployqt`/`appimagetool`/`AppImageUpdate` tooling. @TheAssassin may know the current status and what may be missing to achieve this.

**Comment by danieleds** [comment]
> Continuous builds (pre-release) should update to continuous builds, whereas released (tagged) builds should update to releases (tagged) builds.

Exactly, that's what I had in mind.

If that's not currently possible, it wouldn't be bad as an alternative to make the continuous build update to a released (tagged) build as soon as there is a newer stable release. Basically what I care about is that users of a continuous build never risk staying on that same unstable commit forever.

About Pull Request builds, I guess they could have whatever behavior is easier to implement :)

**Comment by TheAssassin** [comment]
It's all documented: https://github.com/AppImage/AppImageSpec/blob/master/draft.md#github-releases

> Please note that pre-releases are not being considered when using latest. You will have to explicitly provide the name of a release. When using e.g., uploadtool, the name of the release created will always be continuous, hence, you can just specify that value instead of latest.

So, "stable" channel: put in `latest`, which will be resolved automatically to the last non-prerelease. "continuous" channel: use `continuous` instead of `latest`, which will not automatically resolve to the latest pre-release, but statically point to a tag `continuous`.

**Comment by maverick74** [comment]
Sorry for asking but: are we going to have appimage's for notepadqq or not?

It's very useful to have notepadqq on my pocket without having to install it :)

**Comment by danieleds** [comment]
@maverick74 @probonopd @TheAssassin sorry, I was under the assumption that someone else would have been working on it; unfortunately I can't work on it in first person, at this time.

As said, I'll be happy to merge it if someone wants to provide an implementation before I can do it.

**Comment by probonopd** [comment]
Getting

```
Project ERROR: Unknown module(s) in QT: webenginewidgets
make: *** [sub-src-ui-tests-make_first] Error 3
make: *** Waiting for unfinished jobs....
qtchooser: could not find a Qt installation of '5'
Project ERROR: Unknown module(s) in QT: webenginewidgets
make: *** [sub-src-ui-make_first] Error 3
```

Where might it be hiding on https://launchpad.net/~beineri/+archive/ubuntu/opt-qt-5.10.1-trusty?

**Comment by danieleds** [comment]
If it's there, it should be in one of the qt510web*.

I assume qt510webengine.

**Comment by probonopd** [comment]
I am already using that...

**Comment by danieleds** [comment]
Sorry, is that the output of `make`?

What is the output of ./configure?

**Comment by probonopd** [comment]
Here is my build configuration:
https://github.com/probonopd/notepadqq/blob/AppImage/.travis.yml

Here is my full build log:
https://travis-ci.org/probonopd/notepadqq/

Thanks for having a look.

**Comment by JuBan1** [comment]
> sudo apt-get -y install qt510base qt510webchannel qt510websockets qt510svg qt510tools libuchardet-dev

Don't you have to add `qt510webengine`?

**Comment by danieleds** [comment]
> > sudo apt-get -y install qt510base qt510webchannel qt510websockets qt510svg qt510tools libuchardet-dev
> 
> Don't you have to add `qt510webengine`?

that seems to be the problem indeed.

In general, I suggest calling ./configure, so that in case of problems we have an early check. As a reference, this is our travis script for macdeployqt: https://github.com/notepadqq/notepadqq/blob/master/.travis/macos-script.sh

**Comment by danieleds** [comment]
Also note that, if it is more convenient for you, you can also use docker in order to use an environment that is already prepared and works for sure, as we do here: https://github.com/notepadqq/notepadqq/blob/master/.travis/script.sh

I have no idea if it is convenient to use with appimage for bundling graphics libs, though.

**Comment by probonopd** [comment]
My bad, I had confused qt510webengine with qt510webchannel. Never get the difference between the Qt web* projects, and they seem to keep changing stuff...

**Comment by probonopd** [comment]
Do you have any clue what might be going wrong this time? 

Here is my build configuration:
https://github.com/probonopd/notepadqq/blob/AppImage/.travis.yml

Here is my full build log:
https://travis-ci.org/probonopd/notepadqq/

Thanks again.

**Comment by danieleds** [comment]
Unless you have some special need, you should not call qmake directly: please use ./configure as shown in our readme.

**Comment by probonopd** [comment]
New build errors in https://travis-ci.org/probonopd/notepadqq/builds/487815915, which compiler version is required?

**Comment by JuBan1** [comment]
I think you'll need at least GCC 4.9 to get support for that language feature causing this error. But I don't know if 4.9 will be enough as a whole. I's say go GCC 7.x if you can; that ought to be enough.

**Comment by TheAssassin** [comment]
Type inference and lambdas were added in C++11, which was completely implemented in GCC series 4.8 (except for regular expressions which were added in 4.9).

The compiler however is asked to build for C++14 (`-std=gnu++1y`), which isn't available in trusty, please build on xenial.

**Comment by danieleds** [comment]
Either build on xenial as @TheAssassin suggests, or use docker, so you can directly build on a newer Ubuntu where you won't even need to manually add the PPA. I highly encourage this route.

Reference: https://github.com/notepadqq/notepadqq/blob/master/.travis/script.sh

**Comment by probonopd** [comment]
I don't want to build on any distribution newer than the oldest currently still-supported release of Ubuntu, so that the resulting binaries will work on all still-supported distributions of Ubuntu. If building on trusty is not possible, I will wait until trusty is no longer a supported version of Ubuntu.

**Comment by probonopd** [comment]
Seems like my latest build using a newer compiler succeeded, but installation fails with

```
mkdir: cannot create directory ‘/usr/lib/notepadqq/’: Permission denied
make[1]: *** [install_target] Error 1
make[1]: Leaving directory `/home/travis/build/probonopd/notepadqq/src/ui'
make: *** [sub-src-ui-install_subtargets] Error 2
find: `appdir/': No such file or directory
The command "make install DESTDIR=$(readlink -f appdir) ; find appdir/" exited with 1.
```

It seesm to ignore `DESTDIR`. Please advise.

Here is my build configuration:
https://github.com/probonopd/notepadqq/blob/AppImage/.travis.yml

Here is my full build log:
https://travis-ci.org/probonopd/notepadqq/

Thanks again.
